src/obj_char_transfer.c

```diff
@@ -0,0 +1,872 @@
+        if (sObjCharTransferTasksManager->tasks[i].state != OBJ_CHAR_TRANSFER_TASK_INIT) {
+            ObjCharTransferTask_Reset(&sObjCharTransferTasksManager->tasks[i]);
+            --sObjCharTransferTasksManager->num;
+        }
+    }
+}
+
+NNSG2dImageProxy *ObjCharTransfer_GetProxyPtrByResID(int resId) {
+    ObjCharTransferTask *sub = ObjCharTransfer_GetTaskByID(resId);
+    GF_ASSERT(sub != NULL);
+    if (sub->state == OBJ_CHAR_TRANSFER_TASK_INIT) {
+        return NULL;
+    }
+
+    return &sub->imageProxy;
+}
+
+NNSG2dImageProxy *ObjCharTransfer_ResizeTaskByResID_GetProxyPtr(int resId, u32 size) {
+    ObjCharTransferTask *task = ObjCharTransfer_GetTaskByID(resId);
+    GF_ASSERT(task != NULL);
+    if (task->state == OBJ_CHAR_TRANSFER_TASK_INIT) {
+        return NULL;
+    }
+    u32 offsetMain;
+    u32 offsetSub;
+    u32 newSizeMain;
+    u32 newSizeSub;
+    getBlockNumAndFreeSpaceForTransfer(task->vram, &offsetMain, &offsetSub, size, &newSizeMain, &newSizeSub);
+    if (task->state == OBJ_CHAR_TRANSFER_TASK_COPYABLE) {
+        return NULL;
+    }
+    task->state = OBJ_CHAR_TRANSFER_TASK_COPYABLE;
+    ObjCharTransferTask_UpdateBaseAddrs(task, offsetMain, offsetSub);
+    task->needResetBlockTransferFlags = TRUE;
+    task->sizeMain = newSizeMain;
+    task->sizeSub = newSizeSub;
+    ObjCharTransferTask_G2dLoadImageMappingVramTransfer(task);
+    reserveTransferBlocksByVramOffsetAndSize(task->vram, offsetMain, offsetSub, newSizeMain, newSizeSub);
+
+    return &task->imageProxy;
+}
+
+NNSG2dImageProxy *ObjCharTransfer_CopyTransferTaskByProxyPtr(NNSG2dImageProxy *proxyIn) {
+    ObjCharTransferTask *task;
+    ObjCharTransferTask *copiedTask;
+    u32 offsetMain;
+    u32 offsetSub;
+    u32 newSizeMain;
+    u32 newSizeSub;
+    u32 size;
+
+    task = ObjCharTransfer_GetTaskByImageProxy(proxyIn);
+    GF_ASSERT(task != NULL);
+    copiedTask = ObjCharTransfer_GetFreeTask();
+    GF_ASSERT(copiedTask != NULL);
+    if (task->state != OBJ_CHAR_TRANSFER_TASK_COPYABLE) {
+        return NULL;
+    }
+    *copiedTask = *task;
+    copiedTask->state = OBJ_CHAR_TRANSFER_TASK_COPY;
+    if (copiedTask->vram & NNS_G2D_VRAM_TYPE_2DMAIN) {
+        size = copiedTask->sizeMain;
+    } else {
+        size = copiedTask->sizeSub;
+    }
+
+    getBlockNumAndFreeSpaceForTransfer(copiedTask->vram, &offsetMain, &offsetSub, size, &newSizeMain, &newSizeSub);
+    ObjCharTransferTask_UpdateBaseAddrs(copiedTask, offsetMain, offsetSub);
+    copiedTask->needResetBlockTransferFlags = TRUE;
+    copiedTask->sizeMain = newSizeMain;
+    copiedTask->sizeSub = newSizeSub;
+    ObjCharTransferTask_G2dLoadImageMappingVramTransfer(copiedTask);
+    reserveTransferBlocksByVramOffsetAndSize(copiedTask->vram, offsetMain, offsetSub, newSizeMain, newSizeSub);
+
+    return &copiedTask->imageProxy;
+}
+
+void ObjCharTransfer_DeleteTaskCopyByProxyPtr(const NNSG2dImageProxy *proxy) {
+    int i;
+    for (i = 0; i < sObjCharTransferTasksManager->max; ++i) {
```

Comment:
pls declare i within the for loop we use c99 here

Reply:
didn't match

Reply:
weird, don't see why it wouldn't but ig the compiler is a bit fussy sometimes